### PR TITLE
Fix Adam optimizer for half-precision parameters

### PR DESCRIPTION
**Pull request by njroussel**
In the Adam optimizer the EMA debiasing term was explicitly always set to be of type `Float32`. This is an issue when the parameter being optimized is not `Float32` but `Float16` for example, as the learning rate cache checks that its output type matches the parameter's gradient type.

### CODE REVIEW

**Comment by wjakob** [comment]
In general, to train fp16, you will want to keep the whole optimizer state in fp32. I don't think it makes much sense to have a value in fp16 and momentum in fp16. So I am fine with this change.

**Comment by njroussel** [comment]
Ok :+1:  I didn't want to mess with the numerical precision so I had originally made it consistent with the original implementation (prior to regression).

I've now pushed a new commit. The momentum, step and next value are all computed in fp32 (the next value is then downcast to fp16). For double precision parameters, everything will be done in double precision.
